### PR TITLE
Telemetry

### PR DESCRIPTION
**Pull request by ErikSchierboom**
```ruby
ActiveRecord::Base.transaction(isolation: Exercism::READ_COMMITTED) do
  Solution.includes(:track, :user).find_each {|s| Metric::Create.(:submit_solution, s.created_at, track: s.track, user: s.user, solution: s) }
end
ActiveRecord::Base.transaction(isolation: Exercism::READ_COMMITTED) do
  Solution.includes(:track, :user).completed.find_each {|s| Metric::Create.(:complete_solution, s.completed_at, track: s.track, user: s.user, solution: s) }
end
ActiveRecord::Base.transaction(isolation: Exercism::READ_COMMITTED) do
  Solution.includes(:track, :user).published.find_each {|s| Metric::Create.(:publish_solution, s.published_at, track: s.track, user: s.user, solution: s) }
end
ActiveRecord::Base.transaction(isolation: Exercism::READ_COMMITTED) do
  Mentor::Request.includes(:track, :student).where(external: false).find_each {|r| Metric::Create.(:request_mentoring, r.created_at, track: r.track, user: r.student, request: r) }
end
ActiveRecord::Base.transaction(isolation: Exercism::READ_COMMITTED) do
  Mentor::Request.includes(:track, :student).where(external: true).find_each {|r| Metric::Create.(:request_private_mentoring, r.created_at, track: r.track, user: r.student, request: r) }
end
ActiveRecord::Base.transaction(isolation: Exercism::READ_COMMITTED) do
  Mentor::Discussion.finished_for_student.includes(solution: {exercise: :track}).includes(:student).find_each {|d| Metric::Create.(:finish_mentoring, d.finished_at, track: d.track, user: d.student, discussion: d) }
end
ActiveRecord::Base.transaction(isolation: Exercism::READ_COMMITTED) do
  Github::Issue.find_each {|i| Metric::Create.(:open_issue, i.opened_at, track: Track.for_repo(i.repo), user: User.find_by(github_username: i.opened_by_username), issue: i) }
end
ActiveRecord::Base.transaction(isolation: Exercism::READ_COMMITTED) do
  Github::PullRequest.find_each {|pr| Metric::Create.(:open_pull_request, pr.data[:created_at], track: Track.for_repo(pr.repo), user: User.find_by(github_username: User.find_by(github_username: pr.author_username)), pull_request: pr) }
end
ActiveRecord::Base.transaction(isolation: Exercism::READ_COMMITTED) do
  Github::PullRequest.merged.find_each {|pr| Metric::Create.(:merge_pull_request, pr.data[:merged_at], track: Track.for_repo(pr.repo), user: User.find_by(github_username: User.find_by(github_username: pr.merged_by_username)), pull_request: pr) }
end
```

## TODO

- [x] Date range filtering on searching (in the SQL)
- [x] Correct indexes
- [x] Add and use `occurred_at`
- [x] Add uniqueness column on metrics (to stop duplicates)
- [x] Add job every 15mins to recalculate last 60mins of metrics in case Sidekiq fails for any period (this is a terrible version: `60.times.map { |idx| Time.now.to_i - (60 * idx) }.each { |time|MetricPeriod::UpdateMinuteMetrics.(Time.at(time)) }`)
- [x] Use read-committed
- [x] Use find_create_or_find_by!

edit:

```ruby
batch_file = '/tmp/batch.txt'
last_id = File.exists?(batch_file) ? File.read(batch_file).strip.to_i : 0
batch_size = 1000

Solution.includes(:track, :user).where('id > ?', last_id).find_in_batches(batch_size:) do |solutions|
  solutions.each do |s|
    Metric::Create.(:submit_solution, s.created_at, track: s.track, user: s.user, solution: s)
  end

  File.write('/tmp/batch.txt', solutions.last.id)
end
```

### CODE REVIEW

File: app/commands/github/issue/create_or_update.rb [hunk 1]
```diff
@@ -6,7 +6,7 @@ class CreateOrUpdate
       initialize_with :node_id, :attributes
 
       def call
-        issue = ::Github::Issue.create_or_find_by!(node_id:) do |i|
+        @issue = ::Github::Issue.create_or_find_by!(node_id:) do |i|
```
Comment (iHiD) [inline app/commands/github/issue/create_or_update.rb:9]: I'd do `memoize def issue ...` for this.

File: app/commands/github/issue/create_or_update.rb [hunk 2]
```diff
@@ -15,6 +15,8 @@ def call
           i.opened_by_username = attributes[:opened_by_username]
         end
 
+        log_metric! if issue.just_created?
```
Comment (iHiD) [inline app/commands/github/issue/create_or_update.rb]:
```suggestion
        log_metric!(issue) if issue.just_created?
```

File: app/commands/github/issue/create_or_update.rb (hunk 1 above)
Comment (iHiD) [inline app/commands/github/issue/create_or_update.rb]:
```suggestion
        issue = ::Github::Issue.create_or_find_by!(node_id:) do |i|
```

File: app/commands/github/pull_request/create_or_update.rb [hunk 3]
```diff
@@ -25,15 +27,40 @@ def call
...
+
+      def log_open_metrics!
+        LogMetricJob.perform_later(:open_pull_request, created_at, track:, user: author)
+      end
+
+      def log_merge_metrics!
+        LogMetricJob.perform_later(:merge_pull_request, merged_at, track:, user: merged_by)
+      end
+
+      def track = Track.for_repo(pull_request.repo)
```
Comment (iHiD) [inline app/commands/github/pull_request/create_or_update.rb]:
```suggestion
      memoize
      def track = Track.for_repo(pull_request.repo)
```

File: app/models/concerns/has_metric_action.rb [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    enum action: {
+      submit_solution: 0,
+      complete_solution: 1,
+      publish_solution: 2,
+      request_mentoring: 3,
+      finish_mentoring: 4,
+      open_issue: 5,
+      open_pull_request: 6,
+      merge_pull_request: 7
+    }
```
Comment (iHiD) [inline app/models/concerns/has_metric_action.rb]:
```suggestion
    }, _prefix: true
```

File: app/models/concerns/has_metric_action.rb [hunk 5]
```diff
@@ -0,0 +1,18 @@
+module HasMetricAction
+  extend ActiveSupport::Concern
+
+  included do
+    enum action: {
```
Comment (iHiD) [inline app/models/concerns/has_metric_action.rb]:
```suggestion
    enum metric_action: {
```

File: config/initializers/monkey_patching.rb [hunk 6]
```diff
@@ -0,0 +1 @@
+require "time_extensions"
```
Comment (ErikSchierboom) [inline config/initializers/monkey_patching.rb]: Let me know what you think about this: I've created an initializer to monkey patch some methods onto the `Time`/`DateTime`/`TimeWithZone` classes.

File: app/models/track.rb [hunk 7]
```diff
@@ -33,6 +33,11 @@ def self.for!(param)
     find_by!(slug: param)
   end
 
+  def self.for_repo(repo)
+    slug = repo.gsub(/-(test-runner|analyzer|representer)$/, '').split('/').last
+    find_by(slug:)
+  end
```
Comment (ErikSchierboom) [inline app/models/track.rb:39]: I've extracted this code into a helper method as it was now being used in multiple places.

File: db/migrate/20220615130617_create_metrics.rb [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+      t.column :metric_action, :tinyint, null: false, default: 0
+      t.belongs_to :track, null: true
+      t.belongs_to :user, null: true
+
+      t.timestamps
+
+      t.index :created_at
+    end
+
+    %i[month day].each do |period|
```
Comment (ErikSchierboom) [inline db/migrate/20220615130617_create_metrics.rb:17]: It started out with three keywords, but with two it is hardly worth it 🤷

File: db/migrate/20220615130617_create_metrics.rb [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+        t.timestamps
+
+        t.index :created_at
+        t.index [period, :metric_action, :track_id], unique: true, name: 'uniq'
+      end
+    end
+
+    create_table :metric_period_minutes do |t|
+        t.integer :minute, limit: 2, null: false, default: 0
```
Comment (ErikSchierboom) [inline db/migrate/20220615130617_create_metrics.rb:31]: I had to move this out of the above meta-programming bit as the minutes range from 0 to 1439, which doesn't fit a `tinyint`.

File: app/jobs/log_metric_job.rb [hunk 10]
```diff
@@ -0,0 +1,7 @@
+class LogMetricJob < ApplicationJob
+  queue_as :dribble
```
Comment (ErikSchierboom) [inline app/jobs/log_metric_job.rb]: Is this the right queue?

File: app/jobs/update_day_metrics_job.rb [hunk 11]
```diff
@@ -0,0 +1,5 @@
+class UpdateDayMetricsJob < ApplicationJob
+  queue_as :default
```
Comment (ErikSchierboom) [inline app/jobs/update_day_metrics_job.rb]: Is this the right queue?

File: lib/time_extensions.rb [hunk 12]
```diff
@@ -0,0 +1,16 @@
+module TimeExtensions
+  def min_of_day = hour * 60 + min
+  def prev_min = self - 1.minute
+end
```
Comment (ErikSchierboom) [inline lib/time_extensions.rb]: These methods helped cleanup the code quite a bit. Let me know if this is something you think is a good idea.

File: app/jobs/log_metric_job.rb (hunk 10 above)
Comment (iHiD) [reply app/jobs/log_metric_job.rb]: Make a new queue for these.

Comment (iHiD) [reply app/jobs/log_metric_job.rb]: Add add it to the sidekiq config (and anywhere else its used)

File: app/commands/metric_period/update_day_metrics.rb [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+  include Mandate
+
+  def initialize(day_to_update = Time.current.prev_day)
+    @day_to_update = day_to_update.beginning_of_day
+  end
+
+  def call
+    # TODO: add nil track
+
+    metric_actions.product(Track.all).each do |metric_action, track|
```
Comment (iHiD) [inline app/commands/metric_period/update_day_metrics.rb]:
```suggestion
    ActiveRecord::Base.transaction(isolation: Exercism::READ_COMMITTED) 
      generate_data!
    end
  end
    
  def generate_data!
    metric_actions.product(Track.all).each do |metric_action, track|
```

File: app/models/metrics.rb [hunk 14]
```diff
@@ -0,0 +1,5 @@
+module Metrics
+  def self.table_name_prefix
+    "metrics_"
+  end
+end
```
Comment (iHiD) [inline app/models/metrics.rb]: Pretty sure you can delete this file.

File: app/commands/metric_period/search.rb [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+    @metric_periods = @metric_periods.where("#{period_type} >= ?", options[:period_begin].to_i)
+  end
+
+  def filter_period_end!
+    return if options[:period_end].blank?
+
+    @metric_periods = @metric_periods.where("#{period_type} <= ?", options[:period_end].to_i)
+  end
+
+  def filter_metric_type!
```
Comment (ErikSchierboom) [inline app/commands/metric_period/search.rb:45]: I also added the option to filter on metric type

File: config/sidekiq-schedule.yml [hunk 16]
```diff
@@ -65,3 +65,39 @@ nudge_users_to_request_mentoring:
...
+update_last_week_metrics:
+  cron: "0 12 * * *"
+  class: "UpdateLastWeekMetricsJob"
+  queue: metrics
+
+# Every day at noon
+update_last_quarter_metrics:
+  cron: "0 12 * * *"
+  class: "UpdateLastQuarterMetricsJob"
+  queue: metrics
```
Comment (ErikSchierboom) [inline config/sidekiq-schedule.yml:103]: I have no idea what makes sense for these cron values 🤷

File: app/controllers/api/metrics/periodic_controller.rb [hunk 17]
```diff
@@ -0,0 +1,16 @@
+class API::Metrics::PeriodicController < API::BaseController
+  def index
+    period_type = index_params.require(:period_type)
```
Comment (iHiD) [inline app/controllers/api/metrics/periodic_controller.rb]: You probably want to catch this and raise an `InvalidMetricPeriodError`, or not use `require` (ie require will raise a Strong Parameters exception if it's not passsed)

File: app/serializers/serialize_metric_periods.rb [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+    metric_periods.includes(:track).map do |metric_period|
+      serialize_metric_period(metric_period)
+    end
+  end
+
+  private
+  def serialize_metric_period(metric_period)
+    key = metric_period.class.name.demodulize.underscore
+    value = metric_period.send(key)
+    {
```
Comment (iHiD) [inline app/serializers/serialize_metric_periods.rb:17]: Add line-break above this pls.

File: test/controllers/api/metrics/periodic_controller_test.rb [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+  test "index returns 400 when passing unknown metric type" do
+    setup_user
+
+    get api_metrics_periodic_url(
+      period_type: "day",
+      metric_type: "UnknownMetricType"
+    ), headers: @headers, as: :json
+    assert_response 400
+  end
+end
```
Comment (iHiD) [inline test/controllers/api/metrics/periodic_controller_test.rb:55]: Can you add a test for unauthorised too pls (there's lots you can copy from other api controller tests)? And a test for not passing any params.

File: app/commands/metric/queue.rb [hunk 20]
```diff
@@ -0,0 +1,11 @@
+class Metric::Queue
+  include Mandate
+
+  initialize_with :type, :occurred_at, :attributes
+
+  def call
+    LogMetricJob.perform_later(type, occurred_at, **attributes)
+  rescue StandardError
+    # Don't crash if the creation fails, e.g. if sidekiq is down
```
Comment (iHiD) [inline app/commands/metric/queue.rb]:
```suggestion
  rescue StandardError => e
    # Don't crash if the creation fails, e.g. if sidekiq is down
    Bugsnag.notify(e)
```

**Review by iHiD — APPROVED** [review]

**Comment by ErikSchierboom** [comment]
I've finished the backporting and ran the following to check if the counts match:

```ruby
p "#{Solution.where.not(user_id: [1, 720_036]).count} submitted solutions, #{Metrics::SubmitSolutionMetric.count} metrics"
p "#{Solution.completed.where.not(user_id: [1, 720_036]).count} completed solutions, #{Metrics::CompleteSolutionMetric.count} metrics"
p "#{Solution.published.where.not(user_id: [1, 720_036]).count} published solutions, #{Metrics::PublishSolutionMetric.count} metrics"
p "#{Mentor::Request.where(external: false).where.not(student_id: [1, 720_036]).count} mentor requests, #{Metrics::RequestMentoringMetric.count} metrics"
p "#{Mentor::Request.where(external: true).where.not(student_id: [1, 720_036]).count} private mentor requests, #{Metrics::RequestPrivateMentoringMetric.count} metrics"
p "#{Mentor::Discussion.finished_for_student.count} finished mentoring, #{Metrics::FinishMentoringMetric.count} metrics"
p "#{Github::Issue.count} issues opened, #{Metrics::OpenIssueMetric.count} metrics"
p "#{Github::PullRequest.count} prs opened, #{Metrics::OpenPullRequestMetric.count} metrics"
p "#{Github::PullRequest.merged.count} prs merged, #{Metrics::MergePullRequestMetric.count} metrics"
```

Unfortunately, there are some minor deviations:

```shell
"3591518 submitted solutions, 3591715 metrics"
"2756319 completed solutions, 2756519 metrics"
"2072942 published solutions, 2073005 metrics"
"326888 mentor requests, 326863 metrics"
"267295 finished mentoring, 267295 metrics"
"20909 issues opened, 20909 metrics"
"45504 prs opened, 45508 metrics"
"38775 prs merged, 38776 metrics"
```

Nothing major, but still something I'll need to look into.

**Pull request merged**